File: ibm/src/dotnet/Matrix/Matrix.cs
using System;
using System.IO;
using System.Diagnostics;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;

namespace Matrix
{
    public class MatrixDotnet
    {
        public JObject Main(JObject args)
        {
            string machineId = File.ReadAllText("/sys/class/dmi/id/product_uuid");
            string instanceId = File.ReadAllText("/proc/self/cgroup");
            string cpuinfo = File.ReadAllText("/proc/cpuinfo");
            string meminfo = File.ReadAllText("/proc/meminfo");
            string uptime = File.ReadAllText("/proc/uptime");

            int n = 100;

            if(args["n"] != null) {
                bool parseOk = Int32.TryParse(args["n"].ToString(), out n);
                if(!parseOk) {
                    n = 100;
                }
            } else {
                n = 100;
            }

            Stopwatch sw = new Stopwatch();
            sw.Start();
		        int[,] result = matrixCalc(n);
            sw.Stop();
		        
            JObject message = new JObject();
            message.Add("success", new JValue(true));
            JObject payload = new JObject();
            payload.Add("test", new JValue("matrix test"));
            payload.Add("n", new JValue(n));
            payload.Add("time", new JValue(sw.Elapsed.TotalMilliseconds));
            message.Add("payload", payload);
            JObject metrics = new JObject();
            metrics.Add("machineid", new JValue(machineId));
            metrics.Add("instanceid", new JValue(instanceId));
            metrics.Add("cpu", new JValue(cpuinfo));
            metrics.Add("mem", new JValue(meminfo));
            metrics.Add("uptime", new JValue(uptime));
            message.Add("metrics", metrics);

            return (message);
        }

        public static int[,] randomTable(int n) {
            Random rnd = new Random();
            int[,] m = new int[n,n];
            for(int i=0;i<n;i++)
            {
                for(int j=0;j<n;j++)
                {
                    Console.WriteLine("i: " + i);
                    Console.WriteLine("j: " + j);
                    m[i, j]= rnd.Next(1, 100);
                }
            }
            Console.WriteLine("ok");
            return m;
        }

        public static int[,] matrixCalc(int n) {
            int[,] matrixA = randomTable(n);
            Console.WriteLine("1st arrray");
            int[,] matrixB = randomTable(n);
            int[,] matrixMult = new int[n, n];
            
            for (int i = 0; i < matrixA.GetLength(0); i++) {
                for (int j = 0; j < matrixB.GetLength(0); j++) {
                    int sum = 0;
                    for (int k = 0; k < matrixA.GetLength(0); k++) {
                        sum += matrixA[i,k] * matrixB[k,j];
                    }
                    matrixMult[i,j] = sum;
                }
            }

            return matrixMult;
        }
    }
}

File: ibm/src/go/matrix/matrix.go
package main

import (
    "time"
    "math/rand"
    "strconv"
    "io/ioutil"
    "log"
)

func Main(params map[string]interface{}) map[string]interface{} {

    buf1, err := ioutil.ReadFile("/proc/self/cgroup")
	if err != nil {
		log.Fatal(err)
    }
    instance_id := string(buf1)

    buf2, err := ioutil.ReadFile("/proc/cpuinfo")
	if err != nil {
		log.Fatal(err)
    }
    cpuinfo := string(buf2)

    buf3, err := ioutil.ReadFile("/proc/meminfo")
	if err != nil {
		log.Fatal(err)
    }
    meminfo := string(buf3)

    buf4, err := ioutil.ReadFile("/proc/uptime")
	if err != nil {
		log.Fatal(err)
    }
    uptime := string(buf4)

    buf5, err := ioutil.ReadFile("/sys/class/dmi/id/product_uuid")
	if err != nil {
		log.Fatal(err)
    }
    machine_id := string(buf5)

    n := 100

    if _, ok := params["n"]; ok {
        n, err = strconv.Atoi(params["n"].(string))
        if err != nil {
            log.Fatal(err)
        }
    } else {
        n = 100
    }

    start := time.Now()
    matrix(n)
    elapsed := time.Since(start)
    
    msg := map[string]interface{}{}
    msg["success"] = true
    msg["payload"] = map[string]interface{}{}
    msg["payload"].(map[string]interface{})["test"] = "matrix test"
    msg["payload"].(map[string]interface{})["n"] = n
    msg["payload"].(map[string]interface{})["time"] = int64(elapsed / time.Millisecond)
    msg["metrics"] = map[string]string{}
    msg["metrics"].(map[string]string)["machineid"] = machine_id
    msg["metrics"].(map[string]string)["instanceid"] = instance_id
    msg["metrics"].(map[string]string)["cpu"] = cpuinfo
    msg["metrics"].(map[string]string)["mem"] = meminfo
    msg["metrics"].(map[string]string)["uptime"] = uptime

    return msg

 }
	
 func randomTable(n int) [][]int {

    s1 := rand.NewSource(time.Now().UnixNano())
    r1 := rand.New(s1)

    m := make([][]int, n) 
    for i:=0; i<n; i++ {
        m[i] = make([]int, n)
        for j:=0; j<n; j++ {
            m[i][j] = r1.Intn(100)
        }
    }
    return m
}

func matrix(n int) [][]int {
    matrixA := randomTable(n)
    matrixB := randomTable(n)
    matrixMult := make([][]int, n)
 
    for i := 0; i < len(matrixA); i++ {
        matrixMult[i] = make([]int, n)
        for j := 0; j < len(matrixB); j++ {
            sum := 0
            for k := 0; k < len(matrixA); k++ {
                sum += matrixA[i][k] * matrixB[k][j]
            }
            matrixMult[i][j] = sum
        }
    }

    return matrixMult
}
